### PR TITLE
Record parse failure reason and location

### PR DESCRIPTION
**Pull request by LeszekSwirski**
In parse_number_string, if there is a parse error, report the specific
error as one of the values in a new parse_error enum, and update
lastmatch to match the error location. This allows users of the library
to print more helpful error messages for invalid inputs.

### CODE REVIEW

**Comment by LeszekSwirski** [comment]
Happy to discuss alternatives to this, this was just the simplest approach that came to mind and I figured it's easier to make a code-based suggestion than an abstract one.

**Comment by lemire** [comment]
>  I figured it's easier to make a code-based suggestion than an abstract one.

Yep.

**Comment by lemire** [comment]
@LeszekSwirski Can you sync your fork with our main branch. I had to do a forced update to fix problems I created earlier today.

Your PR seems quite reasonable and maybe we want to include it in a release very soon.

**Comment by LeszekSwirski** [comment]
Sure, done.

**Comment by lemire** [comment]
I will push a patch release immediately because I want your fix from an earlier PR. Meanwhile let us turn this PR into at least a minor release.

**Comment by LeszekSwirski** [comment]
Sounds reasonable, it's an observable API change if anyone checks `pnr.lastmatch` instead of `pnr.valid`. I could make it backwards compatible by adding a new `const UC* error_location` if you wanted? No strong opinion from my end.

**Comment by lemire** [comment]
@deadalnix Would you have a look? I think that this PR is a nice step up.

**Comment by deadalnix** [comment]
The obvious thing that jump to my eyes here is the lack of tests :)

The feature looks good, the implementation fairly uncontroversial considering the conditions are already checked for, so for me, as long as there is a good test, I think it should go it.

**Comment by LeszekSwirski** [comment]
Fair comment :smile: I'll add some tests tomorrow.

**Comment by lemire** [comment]
@LeszekSwirski Thanks for adding tests.

File: tests/json_fmt.cpp [hunk 1]
```diff
@@ -12,62 +12,148 @@ int main_readme() {
...
     fast_float::parse_options options{ fast_float::chars_format::json };
-    auto answer = fast_float::from_chars_advanced(input.data(), input.data()+input.size(), result, options);
-    if(answer.ec == std::errc()) { std::cerr << "should have failed\n"; return EXIT_FAILURE; }
+  auto answer = fast_float::from_chars_advanced(
+      input.data(), input.data() + input.size(), result, options);
+  if (answer.ec == std::errc()) {
+    std::cerr << "should have failed\n";
+    return EXIT_FAILURE;
+  }
     return EXIT_SUCCESS;
```
Comment (deadalnix) [inline tests/json_fmt.cpp:34]: You have indentation problems here.

@lemire what do you think about adding a clang-format config file so we can avoid this type of review cycles?

Comment (lemire) [reply tests/json_fmt.cpp:34]: @deadalnix Please review https://github.com/fastfloat/fast_float/pull/257

I don't think we want to consider formatting mistakes as blocking for a PR. We can reformat later.

**Comment by lemire** [comment]
This PR looks good to me. I am inviting further review.

This would be at least a minor release.

**Comment by LeszekSwirski** [comment]
@deadalnix ok now the formatting is fixed (just a "format on save" mess up)

**Comment by lemire** [comment]
> ok now the formatting is fixed (just a "format on save" mess up

Thankfully, we have tools to automatically format. Although, for this project, we don't have a standard yet. If you are interested, I have a PR outstanding that would prescribe a given style: 

https://github.com/fastfloat/fast_float/pull/257

**Comment by lemire** [comment]
The tests pass. We will merge this PR unless someone objects in the near future.

**Comment by lemire** [comment]
Merging. This will be part of the next release.

**Pull request merged**